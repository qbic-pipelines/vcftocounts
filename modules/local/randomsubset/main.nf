process RANDOMSUBSET {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/5a/5acacb55c52bec97c61fd34ffa8721fce82ce823005793592e2a80bf71632cd0/data':
        'community.wave.seqera.io/library/bcftools:1.21--4335bec1d7b44d11' }"

    input:
    tuple val(meta), path(vcf), path(index)
    val(fraction)

    output:
    tuple val(meta), path("*.vcf.gz"), emit: vcf
    tuple val(meta), path("*.tbi")   , emit: tbi
    path "versions.yml"              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    randomsubset.sh ${vcf} ${prefix}.vcf ${fraction}

    bgzip ${prefix}.vcf
    tabix -p vcf ${prefix}.vcf.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo | gzip > ${prefix}.vcf.gz
    touch ${prefix}.tbi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bcftools: \$(bcftools --version 2>&1 | head -n1 | sed 's/^.*bcftools //; s/ .*\$//')
    END_VERSIONS
    """
}
