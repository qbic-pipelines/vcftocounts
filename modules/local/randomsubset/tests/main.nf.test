nextflow_process {

    name "Test Process RANDOMSUBSET"
    script "../main.nf"
    process "RANDOMSUBSET"

    tag "modules"
    tag "modules_"
    tag "randomsubset"

    test("sarscov2 - [vcf, tbi]") {

        when {
            process {
                """
                // The input VCF has 9 records so we expect 4 records in the output VCF
                input[0] = [
                    [ id:'out', single_end:false ], // meta map
                    file('https://github.com/nf-core/test-datasets/raw/refs/heads/modules/data/genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                    file('https://github.com/nf-core/test-datasets/raw/refs/heads/modules/data/genomics/sarscov2/illumina/vcf/test.vcf.gz.tbi', checkIfExists: true)
                ]
                input[1] = 0.5
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.vcf.get(0).get(1)).vcf.summary,
                    file(process.out.tbi.get(0).get(1)).name,
                    process.out.versions
                ).match() },
            )
        }

    }

    test("sarscov2 - [vcf, tbi] - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'out', single_end:false ], // meta map
                    file('https://github.com/nf-core/test-datasets/raw/refs/heads/modules/data/genomics/sarscov2/illumina/vcf/test.vcf.gz', checkIfExists: true),
                    file('https://github.com/nf-core/test-datasets/raw/refs/heads/modules/data/genomics/sarscov2/illumina/vcf/test.vcf.gz.tbi', checkIfExists: true)
                ]
                input[1] = 0.00175
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }

    }
}

