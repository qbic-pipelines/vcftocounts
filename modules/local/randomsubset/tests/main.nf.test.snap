{
    "sarscov2 - [vcf, tbi] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "out",
                            "single_end": false
                        },
                        "out.subset.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "out",
                            "single_end": false
                        },
                        "out.subset.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,ee7626565a01c36b7fb7a05f41e0653e"
                ],
                "tbi": [
                    [
                        {
                            "id": "out",
                            "single_end": false
                        },
                        "out.subset.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "out",
                            "single_end": false
                        },
                        "out.subset.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,ee7626565a01c36b7fb7a05f41e0653e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-18T10:53:32.966380045"
    },
    "sarscov2 - [vcf, tbi]": {
        "content": [
            "VcfFile [chromosomes=[MT192765.1], sampleCount=1, variantCount=4, phased=false, phasedAutodetect=false]",
            "out.subset.vcf.gz.tbi",
            [
                "versions.yml:md5,ee7626565a01c36b7fb7a05f41e0653e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-06-18T10:53:26.441286474"
    }
}